File: mls-sim/storage.py
"""MLS 存档持久化 - 将玩家存档保存到本地 JSON 文件"""

import json
import os
import time

ARCHIVE_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "archives")


def _ensure_dir():
    os.makedirs(ARCHIVE_DIR, exist_ok=True)


def _archive_path(script_dir: str) -> str:
    safe_name = os.path.basename(os.path.normpath(script_dir))
    return os.path.join(ARCHIVE_DIR, f"{safe_name}.json")


def save_room_archives(room):
    """房间停止时持久化所有玩家存档"""
    _ensure_dir()
    path = _archive_path(room.script_dir)

    existing = {}
    if os.path.exists(path):
        try:
            with open(path, "r", encoding="utf-8") as f:
                existing = json.load(f)
        except (json.JSONDecodeError, OSError):
            existing = {}

    for idx, p in room.players.items():
        key = str(idx)
        existing[key] = {
            "name": p.name,
            "map_level": p.map_level,
            "map_exp": p.map_exp,
            "played_count": p.played_count,
            "items": dict(p.items),
            "script_archive": p.script_archive,
            "common_archive": dict(p.common_archive),
            "read_archive": dict(p.read_archive),
            "cfg_archive": dict(p.cfg_archive),
            "saved_at": time.strftime("%Y-%m-%d %H:%M:%S"),
        }

    with open(path, "w", encoding="utf-8") as f:
        json.dump(existing, f, ensure_ascii=False, indent=2)

    return path


def load_player_archives(script_dir: str) -> dict:
    """加载指定脚本目录的存档数据，返回 {player_index_str: archive_dict}"""
    path = _archive_path(script_dir)
    if not os.path.exists(path):
        return {}
    try:
        with open(path, "r", encoding="utf-8") as f:
            return json.load(f)
    except (json.JSONDecodeError, OSError):
        return {}


def list_archives() -> list[dict]:
    """列出所有已保存的存档"""
    _ensure_dir()
    result = []
    for fname in os.listdir(ARCHIVE_DIR):
        if fname.endswith(".json"):
            path = os.path.join(ARCHIVE_DIR, fname)
            try:
                with open(path, "r", encoding="utf-8") as f:
                    data = json.load(f)
                result.append({
                    "script": fname[:-5],
                    "players": len(data),
                    "file": path,
                })
            except (json.JSONDecodeError, OSError):
                continue
    return result

File: mls-sim-rs/src/storage.rs
use serde_json::Value;
use std::collections::HashMap;
use std::path::{Path, PathBuf};

pub fn archive_dir(base: &str) -> PathBuf {
    let p = PathBuf::from(base);
    std::fs::create_dir_all(&p).ok();
    p
}

fn archive_path(base: &str, script_dir: &str) -> PathBuf {
    let name = Path::new(script_dir)
        .file_name()
        .map(|n| n.to_string_lossy().to_string())
        .unwrap_or_else(|| "unknown".into());
    archive_dir(base).join(format!("{}.json", name))
}

pub fn save_room_archives(
    base: &str,
    script_dir: &str,
    players: &HashMap<i32, crate::player::Player>,
) -> std::io::Result<PathBuf> {
    let path = archive_path(base, script_dir);
    let mut existing: serde_json::Map<String, Value> = if path.exists() {
        match std::fs::read_to_string(&path) {
            Ok(text) => serde_json::from_str(&text).unwrap_or_default(),
            Err(_) => Default::default(),
        }
    } else {
        Default::default()
    };

    for (idx, p) in players {
        let now = chrono::Local::now().format("%Y-%m-%d %H:%M:%S").to_string();
        existing.insert(
            idx.to_string(),
            serde_json::json!({
                "name": p.name,
                "map_level": p.map_level,
                "map_exp": p.map_exp,
                "played_count": p.played_count,
                "items": p.items,
                "script_archive": p.script_archive,
                "common_archive": p.common_archive,
                "read_archive": p.read_archive,
                "cfg_archive": p.cfg_archive,
                "saved_at": now,
            }),
        );
    }

    let json = serde_json::to_string_pretty(&existing)?;
    std::fs::write(&path, json)?;
    Ok(path)
}

#[allow(dead_code)]
pub fn load_player_archives(base: &str, script_name: &str) -> Value {
    let path = archive_dir(base).join(format!("{}.json", script_name));
    if !path.exists() {
        return Value::Object(Default::default());
    }
    match std::fs::read_to_string(&path) {
        Ok(text) => serde_json::from_str(&text).unwrap_or(Value::Object(Default::default())),
        Err(_) => Value::Object(Default::default()),
    }
}

pub fn apply_saved_archives(
    base: &str,
    script_dir: &str,
    players: &mut HashMap<i32, crate::player::Player>,
) {
    let name = Path::new(script_dir)
        .file_name()
        .map(|n| n.to_string_lossy().to_string())
        .unwrap_or_else(|| "unknown".into());
    let path = archive_dir(base).join(format!("{}.json", name));
    if !path.exists() {
        return;
    }
    let data: serde_json::Map<String, Value> = match std::fs::read_to_string(&path) {
        Ok(text) => serde_json::from_str(&text).unwrap_or_default(),
        Err(_) => return,
    };

    for (idx, player) in players.iter_mut() {
        let key = idx.to_string();
        if let Some(saved) = data.get(&key) {
            if let Some(v) = saved.get("script_archive").and_then(|v| v.as_str()) {
                if player.script_archive.is_none() {
                    player.script_archive = Some(v.to_string());
                }
            }
            if let Some(obj) = saved.get("common_archive").and_then(|v| v.as_object()) {
                if player.common_archive.is_empty() {
                    player.common_archive = obj
                        .iter()
                        .map(|(k, v)| (k.clone(), v.as_str().unwrap_or("").to_string()))
                        .collect();
                }
            }
            if let Some(obj) = saved.get("read_archive").and_then(|v| v.as_object()) {
                if player.read_archive.is_empty() {
                    player.read_archive = obj
                        .iter()
                        .map(|(k, v)| (k.clone(), v.as_str().unwrap_or("").to_string()))
                        .collect();
                }
            }
            if let Some(obj) = saved.get("cfg_archive").and_then(|v| v.as_object()) {
                if player.cfg_archive.is_empty() {
                    player.cfg_archive = obj
                        .iter()
                        .map(|(k, v)| (k.clone(), v.as_str().unwrap_or("").to_string()))
                        .collect();
                }
            }
            if let Some(obj) = saved.get("items").and_then(|v| v.as_object()) {
                if player.items.is_empty() {
                    player.items = obj
                        .iter()
                        .map(|(k, v)| (k.clone(), v.as_i64().unwrap_or(0) as i32))
                        .collect();
                }
            }
            if let Some(v) = saved.get("map_level").and_then(|v| v.as_i64()) {
                if player.map_level == 1 {
                    player.map_level = v as i32;
                }
            }
            if let Some(v) = saved.get("map_exp").and_then(|v| v.as_i64()) {
                if player.map_exp == 0 {
                    player.map_exp = v as i32;
                }
            }
            if let Some(v) = saved.get("played_count").and_then(|v| v.as_i64()) {
                if player.played_count == 0 {
                    player.played_count = v as i32;
                }
            }
        }
    }
}

#[allow(dead_code)]
pub fn list_archives(base: &str) -> Vec<Value> {
    let dir = archive_dir(base);
    let mut result = Vec::new();
    if let Ok(entries) = std::fs::read_dir(&dir) {
        for entry in entries.flatten() {
            let fname = entry.file_name().to_string_lossy().to_string();
            if fname.ends_with(".json") {
                if let Ok(text) = std::fs::read_to_string(entry.path()) {
                    if let Ok(data) = serde_json::from_str::<serde_json::Map<String, Value>>(&text)
                    {
                        result.push(serde_json::json!({
                            "script": &fname[..fname.len()-5],
                            "players": data.len(),
                            "file": entry.path().to_string_lossy(),
                        }));
                    }
                }
            }
        }
    }
    result
}
